### PR TITLE
Module loaders

### LINKED ISSUE #494: Module loaders work
Our final solution should meet the following criteria:
- We should have the ability to use ES6 syntax in all of our app files
- All of our app files should be imported and not use global variables
  - Forces us to decouple our modules and stick to consistent design patterns
  - Allows us to eliminate dead code
- All dependencies should be imported where used, and not assigned to global variables
  - Gives visibility into where dependencies are used
Allows us to remove or replaces dependencies when needed
- All scripts should be bundled into "app" and "vendor" bundles
  - Reduces the number of HTTP requests
  - Allows us to cache dependencies and app code separately
  - Allows us to apply minification and other transforms to all scripts
- We should have a manifest of all of our dependencies in package.json
  - Provides a record for ourselves and contributors to know which dependencies are being used
  - Allows us to use tools to analyze and optimize our dependency graph
  - Allows us to apply security updates, bug fixes and new features from our dependencies quickly
  - Reduces the amount of third party code committed to our repo

### LINKED ISSUE #528: Large "+" buttons in Data Download on initial load [dd.sel.1]
When zoomed out, the data download buttons do no scale properly. 
[Issue Example](https://worldview.earthdata.nasa.gov/?p=geographic&l=MODIS_Aqua_CorrectedReflectance_TrueColor(hidden),MODIS_Terra_CorrectedReflectance_TrueColor,MODIS_Terra_Aerosol,Coastlines&t=2013-09-28&z=3&v=-347.0625,-226.40625,347.0625,226.40625&ab=off&as=2013-09-28&ae=2013-10-05&av=3&al=true&download=MOD04_L2)
Internal Test `dd.sel.1`. 
Affects IE11, Chrome, FF. Edge works as expected.

### LINKED ISSUE #673: Update current year references
Any references to 2017 that were meant to indicate the current year should be updated to 2018

### PR DESCRIPTION
**Pull request by Benjaki2**
Resolves #494 
Fixes #528
Fixes #673 

Depends on https://github.com/nasa-gibs/worldview-components/pull/32

# Outstanding Items

## Features to be modularized and re-enabled
- [x] Core
- [x] Tour - @ZachTRice #639
- [x] Animation - @Benjaki2 #640 
- [x] Data Download - @Benjaki2 #654 
- [x] Image Download - @Benjaki2 #645
- [x] Natural Events - @localjo #648 
- [x] Notifications - @Benjaki2 #644
- [x] URL Shortening (done along with other features)
- [x] Custom Palettes  (done along with other features)
- [x] What's New - @Benjaki2 (#644)
- [x] Feedback (done along with #639)
- [x] Arctic Projection Change - @localjo #650 
- [x] Google Analytics (done along with other features)
- [x] Alert (done along with other features)
- [x] Timeline mobile wheels - @localjo #651 
- [x] Debug.js - @localjo #653 

## Update Buster Tests 
All tests should be modularized, and the code should be tested against them before that section is marked complete)
- [x] Core files (fixtures.js, wv.brand-test.js)
- [x] Data
- [x] Date
- [x] Layer
- [x] Link
- [x] Map
- [x] Palettes
- [x] Proj
- [x] UI
- [x] Util

## Other Tasks
- [x] Clean up build scripts
- [x] Remove converted dependencies (wv.js.json, dependencies, etc.)
- [x] Move all dependencies from `/ext` to package.json
- [x] Make sure all required CSS files are included (some have been removed)
- [x] Shim or include jscache library in package.json
- [x] Enable watch mode for JS and CSS build tasks - @localjo #647 
- [x] Make build script work for dev and production environments - @localjo #647 

## Bugs
- [x] Fix projection button (Getting the error "Cannot read property 'close' of undefined in projection/projection/ui.js")
- [x] Fix "Add layer" errors (in layer module)
- [x] Fix error related to WVC on tab switching
- [x] The URL state is not updating
- [x] iCheck boxes are missing `iCheck iCheck-checkbox` classes on div which is affecting the styling
- [x] Hovering a layer on the map the includes a colorbar does not update the label in the sidebar. (`.wv-running` / `.wv-active` classes not being added from `map/runningdata.js`)
- [x] `jQuery.easing[this.easing] is not a function` error when opening projection switcher.
- [x] Styling of projection dropdown is off (more spacing and not highlighted)
- [x] `React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components).` when trying to open share link button.
- [x] jQuery UI default stylesheet changing causes a number of elements to appear different.
-- Text appears darker than before. This can be seen in the Events / Date tab, Measurement setting tab (Add Layers modal), 
-- The "Squash Layer" checkbox in the sidebar has no border and is missing padding/margin right
-- The radio "check" buttons under Color palettes in the sidebar is missing thick borders
-- jQuery UI default font-stack is no longer `Segoe UI,Arial,sans-serif`, it is `Verdana,Arial,sans-serif`, so the "Add Layers" button looks different
- [x] Add Layers -> Measurements description drawer button not working (no error displayed) 
Note* This is related to toggle-event being removed in jQuery 1.9+ (https://api.jquery.com/toggle-event/)
- [x] Map Scale Bug 

## Cleanup
- [x] Final pass over `main.js` @localjo #656 
- [x] Make sure all lodash variables are using the same naming convention
- [x] Review usage of `NODE_ENV` in production  - @localjo #659 
- [x] Make sure build script changes work with GIBS options repo - @localjo #659
- [x] Move all build scripts into a single directory (instead of root, `/bin`, `/tasks`, and `/deploy`) - @localjo #664 
- [x] Run e2e tests to confirm functionality
- [x] Run test deployment on SIT/UAT (May need to coordinate with Jon)  - @localjo #664 
- [x] Update `worldview-private` deployment scripts - @localjo 
- [x] Fix Babel not compiling WVC code
- [x] Go through manual test plan
- [x] Fix any bugs found (see [bugs attached to Milestone](https://github.com/nasa-gibs/worldview/milestone/16))
- [x] Review all documentation to make sure it's up to date
- [ ] Release https://github.com/nasa-gibs/worldview-components/pull/32 and update package.json

### CODE REVIEW

**Comment by Benjaki2** [comment]
Note 2** there are onload functions in `wv.util.browser` to be reworked into the onload function in `wv.main.js`

**Comment by localjo** [comment]
Getting this message on `npm install`;

> We're super 😸  excited that you're trying to use ES2015 syntax, but instead of making more yearly presets 😭 , Babel now has a better preset that we recommend you use instead: npm install babel-preset-env --save-dev. preset-env without options will compile ES2015+ down to ES5 just like using all the presets together and thus is more future proof. It also allows you to target specific browsers so that Babel can do less work and you can ship native ES2015+ to user 😎 ! We are also in the process of releasing v7, so please give http://babeljs.io/blog/2017/09/12/planning-for-7.0 a read and help test it out in beta! Thanks so much for using Babel 🙏, please give us a follow on Twitter @babeljs for news on Babel, join slack.babeljs.io for discussion/development and help support the project at opencollective.com/babel

**Comment by Benjaki2** [comment]
@localjo Yeah I saw that yesterday. Need to switch the `preset-package-name`

File: package.json [hunk 1]
```diff
@@ -20,11 +20,16 @@
     "e2e:chrome": "nightwatch -e chrome",
     "e2e:firefox": "nightwatch -e firefox --skiptags wip",
     "browserstack": "node e2e/browserstack.js -c e2e/browserstack.conf.js -e browserstack --skiptags wip",
-    "report": "node e2e/create-report.js"
+    "report": "node e2e/create-report.js",
+    "build-js": "node ./tasks/build-js"
```
Comment (localjo) [inline package.json]: Can we make this `build`?

File: package.json [hunk 2]
```diff
@@ -76,12 +83,85 @@
...
+    "proj4": "2.3.3",
     "promise-queue": "2.2.3",
     "react": "15.3.1",
     "react-dom": "15.3.1",
-    "worldview-components": "^1.5.1",
-    "worldview-options-eosdis": "nasa-gibs/worldview-options-eosdis#v2.1.2"
+    "worldview-components": "^1.5.2",
+    "worldview-options-eosdis": "github:nasa-gibs/worldview-options-eosdis#v2.1.2"
+  },
+  "browser": {
```
Comment (localjo) [inline package.json]: @Benjaki2 What's this browser key? How does it work? These dependencies should be installed in `dependencies` via npm, not in the `web/ext` folder.

Comment (localjo) [reply package.json]: I found the documentation for the browser key: https://github.com/browserify/browserify#browser-field

But we should update this to use package names, and make sure that those packages are in the dependencies field.

Comment (localjo) [reply package.json]: I did this here: https://github.com/nasa-gibs/worldview/pull/613

**Comment by localjo** [comment]
Did we make changes to the Gruntfile and package-lock.json in `development`? Where did these new conflicts come from?

**Comment by localjo** [comment]
I don't see any changes on `development`, and the conflicts that I see when I try to resolve are only changes on the `module-loaders` branch. I'm not sure why git/GitHub are suddenly thinking there are conflicts here.

**Comment by ZachTRice** [comment]
After pulling in the latest version of module-loaders into the layers fix branch, it looks like some of the CSS involving the check boxes was affected. See:
screenshot_10

**Comment by localjo** [comment]
We might be missing some of the CSS files that were included before. We'll have to investigate.

**Comment by localjo** [comment]
I'm seeing no map and the following error when I do a fresh install of this branch; `TypeError: _react2.default is undefined`.

**Comment by localjo** [comment]
Seems like that error is related to `worldview-components`. In Chrome the error I'm seeing, on the same line, is `Uncaught TypeError: Cannot read property 'PropTypes' of undefined`. And the stack trace points to the `react-input-range` reference in ` worldview-components/src/scripts/components/animationwidget/wv.animation.widget.js`. To reproduce the error, try re-installing and re-linking the latest `worldview-components`.

**Comment by localjo** [comment]
Looks like the problem was that I forgot to use the right version of `worldview-components`: https://github.com/nasa-gibs/worldview-components/pull/32

**Comment by ZachTRice** [comment]
During cleanup, let's discuss changing the `NODE_ENV` check to `if (process.env.NODE_ENV !== 'production')`. Either way, we should have `production` set on Bamboo for performance benefits.

**Comment by localjo** [comment]
Seeing a bug with the map scale:

screen shot 2018-01-04 at 4 57 23 pm

**Comment by ZachTRice** [comment]
Making a note about timewheel; I was seeing a very minor error related to `classList equals null`. When we do the once-over run, let's keep an eye on this in case it leads to more serious issues.

**Comment by localjo** [comment]
Fixing sourcemaps in and moving build scripts is much more complex than I had hoped, so let's circle back around and address those things later as separate issues so we can get this PR tested and merged by our original deadline.

**Comment by localjo** [comment]
The latest on this branch has been deployed to http://worldview.sit.earthdata.nasa.gov and is ready for a final run-through and, assuming all goes well, then deployment.

**Comment by localjo** [comment]
Before merging, we do need to comb through the docs and update things that have changed.

**Pull request merged**